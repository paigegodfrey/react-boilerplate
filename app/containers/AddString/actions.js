/*
 *
 * AddString actions
 *
 */

import {
  CHANGE_STRING,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
  RESET_FORM,
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {string} newString The new text of the input field
 *
 * @return {object} An action object with a type of CHANGE_STRING
 */
export function changeString(newString) {
  return {
    type: CHANGE_STRING,
    newString,
  };
}

/**
 * Add new string, this action starts the request saga
 *
 * @return {object} An action object with a type of ADD_STRING
 */
export function addString() {
  return {
    type: ADD_STRING,
  };
}

/**
 * Dispatched when the new string is added by the request saga
 *
 * @param  {string} newString The new string
 *
 * @return {object}      An action object with a type of ADD_STRING_SUCCESS passing the string
 */
export function stringAdded(newString) {
  return {
    type: ADD_STRING_SUCCESS,
    newString,
  };
}

/**
 * Dispatched when adding the strings fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of ADD_STRING_ERROR passing the error
 */
export function stringAddingError(error) {
  return {
    type: ADD_STRING_ERROR,
    error,
  };
}

/**
 * Dispatched when form submission is complete
 *
 * @return {object}       An action object with a type of RESET_FORM
 */
export function resetForm() {
  return {
    type: RESET_FORM,
  };
}
