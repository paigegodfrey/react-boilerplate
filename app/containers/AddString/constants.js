/*
 *
 * AddString constants
 *
 */

export const DEFAULT_ACTION = 'app/AddString/DEFAULT_ACTION';
