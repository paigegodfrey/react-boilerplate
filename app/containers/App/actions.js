/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_STRINGS,
  LOAD_STRINGS_SUCCESS,
  LOAD_STRINGS_ERROR,
  ADD_STRING,
  ADD_STRING_SUCCESS,
  ADD_STRING_ERROR,
} from './constants';

/**
 * Load the strings, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_STRINGS
 */
export function loadStrings() {
  return {
    type: LOAD_STRINGS,
  };
}

/**
 * Dispatched when the strings are loaded by the request saga
 *
 * @param  {array} strings The string data
 *
 * @return {object}      An action object with a type of LOAD_STRINGS_SUCCESS passing the strings
 */
export function stringsLoaded(strings) {
  return {
    type: LOAD_STRINGS_SUCCESS,
    strings,
  };
}

/**
 * Dispatched when loading the strings fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_STRINGS_ERROR passing the error
 */
export function stringLoadingError(error) {
  return {
    type: LOAD_STRINGS_ERROR,
    error,
  };
}

/**
 * Add new string, this action starts the request saga
 *
 * @return {object} An action object with a type of ADD_STRING
 */
export function addString() {
  return {
    type: ADD_STRING,
  };
}

/**
 * Dispatched when the new string is added by the request saga
 *
 * @param  {array} string The new string data
 *
 * @return {object}      An action object with a type of ADD_STRING_SUCCESS passing the strings
 */
export function stringAdded(string) {
  return {
    type: ADD_STRING_SUCCESS,
    string,
  };
}

/**
 * Dispatched when loading the strings fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of ADD_STRING_ERROR passing the error
 */
export function stringAddingError(error) {
  return {
    type: ADD_STRING_ERROR,
    error,
  };
}
