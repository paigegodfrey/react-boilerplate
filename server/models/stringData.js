let data = ['Original string!', 'Second string!', 'Third string!'];

// eslint-disable-next-line no-unused-vars
class StringData {
  static async findAll() {
    if (data.length === 0) {
      const error = new Error('No strings available');
      error.status = 202;
      throw error;
    } else {
      return data;
    }
  }

  static async create(newString) {
    data = [newString, ...data];
    return newString;
  }
}

module.exports = StringData;
