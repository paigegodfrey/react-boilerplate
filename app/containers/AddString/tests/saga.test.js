/**
 * Test sagas
 */

import { put, takeLatest } from 'redux-saga/effects';
import { ADD_STRING } from '../constants';
import { stringAdded, stringAddingError } from '../actions';
import stringData, { postString } from '../saga';

const newString = 'test string';

describe('addStringSaga Saga', () => {
  let addStringGenerator;

  beforeEach(() => {
    addStringGenerator = postString();

    const selectDescriptor = addStringGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = addStringGenerator.next(newString).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the stringAdded action if it requests the data successfully', () => {
    const response = newString;
    const putDescriptor = addStringGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(stringAdded(response, newString)));
  });

  it('should call the stringAddingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = addStringGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(stringAddingError(response)));
  });
});

describe('stringDataSaga Saga', () => {
  const stringDataSaga = stringData();

  it('should start task to watch for ADD_STRING action', () => {
    const takeLatestDescriptor = stringDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(ADD_STRING, postString));
  });
});
