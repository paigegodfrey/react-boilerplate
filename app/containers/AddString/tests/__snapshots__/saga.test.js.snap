// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addStringSaga Saga should call the stringAddingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addStringSaga Saga should call the stringAddingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`addStringSaga Saga should dispatch the stringAdded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`addStringSaga Saga should dispatch the stringAdded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
